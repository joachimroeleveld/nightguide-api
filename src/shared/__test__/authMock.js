const { TEST_USER_1 } = require('../../shared/__test__/fixtures/userFixtures');
const auth = require('../../framework/auth');
const User = require('../../routes/users/userModel');

function mockUserAuth(sandbox = global.sandbox) {
  const user = new User(TEST_USER_1);
  user.checkRole = sandbox.stub().returns(true);
  sandbox.stub(auth, 'getUserFromToken').resolves(user);
}

function restoreUserAuth() {
  auth.getUserFromToken.restore();
}

function mockAppClientAuth(sandbox = global.sandbox) {
  sandbox.stub(auth, 'checkAppTokenHeader').returns(true);
}

function restoreAppClientAuth() {
  auth.checkAppTokenHeader.restore();
}

module.exports = {
  mockUserAuth,
  restoreUserAuth,
  mockAppClientAuth,
  restoreAppClientAuth,
};
