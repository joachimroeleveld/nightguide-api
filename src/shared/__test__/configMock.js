const config = require('../config');

const mockedEnv = {
  ...process.env,
};

function mockKey(key, value, sandbox = global.sandbox) {
  mockedEnv[key] = value;
  sandbox.stub(config, 'get').callsFake(key => mockedEnv[key]);
}

module.exports = {
  mockKey,
};
