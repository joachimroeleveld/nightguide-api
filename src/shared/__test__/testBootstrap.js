const dotenv = require('dotenv');
dotenv.load();

const sgMail = require('@sendgrid/mail');
const mongoose = require('mongoose');
const config = require('../config');

const sandbox = require('sinon').createSandbox();
const { mockAuth } = require('./authMock');
const configMock = require('./configMock');

const { createExpressApp } = require('../../framework/expressServer');

beforeAll(cb => {
  // Use global sandbox for mocks that you want to persist across multiple test suites
  if (!global.sandbox) {
    global.sandbox = sandbox;
  }

  mockAuth();

  configMock.mockKey('HOST', 'http://localhost:8080');

  // Mock SendGrid
  sandbox.stub(sgMail, 'send').resolves();

  global.app = createExpressApp();

  mongoose
    .connect(
      config.get('MONGO_URI'),
      { useNewUrlParser: true, dbName: 'main-test' }
    )
    .then(() => cb())
    .catch(err => cb(err));
});

afterAll(async () => {
  mongoose.connection.close();

  global.sandbox.restore();
});
