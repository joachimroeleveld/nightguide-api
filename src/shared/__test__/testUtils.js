const mongoose = require('mongoose');
const _ = require('lodash');

const { deepFreeze } = require('../util/objects');

function generateMongoFixture(base, overrides) {
  return deepFreeze({
    ...base,
    _id: new mongoose.Types.ObjectId().toString(),
    ...overrides,
  });
}

async function resetDb() {
  const removals = Object.keys(mongoose.connection.collections).map(
    collection =>
      new Promise(resolve =>
        mongoose.connection.collections[collection].remove(resolve)
      )
  );

  let dropIndices = [];
  // dropIndices = Object.keys(mongoose.connection.collections).map(
  //   collection =>
  //     new Promise(resolve =>
  //       mongoose.connection.collections[collection].dropAllIndexes(resolve)
  //     )
  // );

  await Promise.all(_.flatten([removals, dropIndices]));
}

function setFixtureLocation(fixture, coordinates) {
  return {
    ...fixture,
    location: {
      ...fixture.location,
      coordinates: {
        type: 'Point',
        coordinates: coordinates,
      },
    },
  };
}

module.exports = {
  resetDb,
  setFixtureLocation,
  generateMongoFixture,
};
