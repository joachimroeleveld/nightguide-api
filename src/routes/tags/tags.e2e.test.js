require('../../shared/__test__/testBootstrap');

const request = require('supertest');
const sinon = require('sinon');

const { validator } = require('../../shared/openapi');
const {
  TEST_TAG_1,
  TEST_TAG_2,
  TEST_TAG_3,
} = require('../../shared/__test__/fixtures');
const { resetDb } = require('../../shared/__test__/testUtils');
const tagRepository = require('./tagRepository');

const TAG_SNAPSHOT_MATCHER = {
  id: expect.any(String),
  createdAt: expect.any(String),
  updatedAt: expect.any(String),
};

const sandbox = sinon.createSandbox();

describe('tags e2e', () => {
  beforeEach(async () => {
    sandbox.restore();
    await resetDb();
  });

  describe('GET /tags', () => {
    const validateResponse = validator.validateResponse('get', '/tags');

    it('happy path', async () => {
      await tagRepository.createTag(TEST_TAG_1);

      const res = await request(global.app).get('/tags');

      expect(res.status).toEqual(200);
      expect(res.body.totalCount).toEqual(1);
      expect(res.body.results[0]).toMatchInlineSnapshot(
        TAG_SNAPSHOT_MATCHER,
        `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": Object {
    "en": "LGBGTQ",
  },
  "slug": "lgbtq",
  "updatedAt": Any<String>,
}
`
      );
      expect(validateResponse(res)).toBeUndefined();
    });

    it('ids filter', async () => {
      const tag1 = await tagRepository.createTag(TEST_TAG_1);
      const tag2 = await tagRepository.createTag(TEST_TAG_2);
      await tagRepository.createTag(TEST_TAG_3);

      const ids = [tag1._id.toString(), tag2._id.toString()];

      const res = await request(global.app)
        .get('/tags')
        .query({
          ids,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(2);
      expect(res.body.results.map(item => item.id)).toEqual(ids);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('slugs filter', async () => {
      const tag1 = await tagRepository.createTag(TEST_TAG_1);
      const tag2 = await tagRepository.createTag(TEST_TAG_2);
      await tagRepository.createTag(TEST_TAG_3);

      const slugs = [tag1.slug, tag2.slug];

      const res = await request(global.app)
        .get('/tags')
        .query({
          slugs,
        });

      expect(res.status).toEqual(200);
      expect(res.body.results.length).toBe(2);
      expect(res.body.results.map(item => item.slug)).toEqual(slugs);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('GET /tags/:tagId', () => {
    const validateResponse = validator.validateResponse('get', '/tags/{tagId}');

    it('happy path', async () => {
      const tag1 = await tagRepository.createTag(TEST_TAG_1);

      const res = await request(global.app)
        .get(`/tags/${tag1._id}`)
        .send()
        .expect(200);

      expect(res.body).toMatchSnapshot(TAG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });

    it('simple fields', async () => {
      const tag1 = await tagRepository.createTag({
        ...TEST_TAG_1,
        slug: 'techno',
        name: {
          en: 'Techno',
        },
      });

      const res = await request(global.app)
        .get(`/tags/${tag1._id}`)
        .send()
        .expect(200);

      expect(res.body).toMatchSnapshot(TAG_SNAPSHOT_MATCHER);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('POST /tags', () => {
    const validateResponse = validator.validateResponse('post', '/tags');

    it('happy path', async () => {
      const res = await request(global.app)
        .post('/tags')
        .send(tagRepository.deserialize(TEST_TAG_1));
      const body = res.body;

      expect(res.status).toEqual(201);
      expect(body.id).toEqual(TEST_TAG_1._id);
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('PUT /tags', () => {
    const validateResponse = validator.validateResponse('put', '/tags/{tagId}');

    it('happy path', async () => {
      const tag1 = await tagRepository.createTag(TEST_TAG_1);

      const res = await request(global.app)
        .put(`/tags/${tag1._id}`)
        .send({
          name: {
            en: 'changed',
          },
        });
      const body = res.body;

      expect(res.status).toEqual(200);
      expect(body.id).toEqual(TEST_TAG_1._id);
      expect(body.name.en).toEqual('changed');
      expect(validateResponse(res)).toBeUndefined();
    });
  });

  describe('DELETE /tags/:tagId', () => {
    const validateResponse = validator.validateResponse(
      'delete',
      '/tags/{tagId}'
    );

    it('happy path', async () => {
      const createdTag = await tagRepository.createTag(TEST_TAG_1);

      const res = await request(global.app).delete(`/tags/${createdTag._id}`);

      let tag = await tagRepository.getTag(createdTag._id);

      expect(tag).toBe(null);
      expect(res.status).toEqual(200);
      expect(validateResponse(res)).toBeUndefined();
    });
  });
});
