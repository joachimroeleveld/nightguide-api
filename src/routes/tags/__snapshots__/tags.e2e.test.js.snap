// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tags e2e GET /tags/:tagId happy path 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "updatedAt": Any<String>,
}
`;

exports[`tags e2e GET /tags/:tagId simple fields 1`] = `
Object {
  "__v": 0,
  "createdAt": Any<String>,
  "id": Any<String>,
  "name": Object {
    "en": "Techno",
  },
  "updatedAt": Any<String>,
}
`;
